The caller function works for this specific situation but fails when an object is passed to the function that is not of the Derived type. When an object is passed that is not of the Derived type the dynamic_cast throws a bad_cast exception.